File: ontrack-extension-git/src/main/java/net/nemerosa/ontrack/extension/git/graphql/GitChangeLogGQLType.java
package net.nemerosa.ontrack.extension.git.graphql;

import graphql.schema.GraphQLObjectType;
import graphql.schema.GraphQLTypeReference;
import net.nemerosa.ontrack.extension.git.model.GitChangeLog;
import net.nemerosa.ontrack.extension.git.service.GitService;
import net.nemerosa.ontrack.graphql.schema.GQLType;
import net.nemerosa.ontrack.graphql.support.GraphqlUtils;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Component;

import static net.nemerosa.ontrack.graphql.support.GraphqlUtils.fetcher;

/**
 * @see net.nemerosa.ontrack.extension.git.model.GitChangeLog
 */
@Component
public class GitChangeLogGQLType implements GQLType {

    public static final String GIT_CHANGE_LOG = "GitChangeLog";

    private final GitUICommitGQLType gitUICommitGQLType;
    private final GitService gitService;

    @Autowired
    public GitChangeLogGQLType(GitUICommitGQLType gitUICommitGQLType, GitService gitService) {
        this.gitUICommitGQLType = gitUICommitGQLType;
        this.gitService = gitService;
    }

    @Override
    public GraphQLTypeReference getTypeRef() {
        return new GraphQLTypeReference(GIT_CHANGE_LOG);
    }

    @Override
    public GraphQLObjectType createType() {
        return GraphQLObjectType.newObject()
                .name(GIT_CHANGE_LOG)
                // Commits
                .field(f -> f.name("commits")
                        .description("List of commits in the change log")
                        .type(GraphqlUtils.stdList(gitUICommitGQLType.getTypeRef()))
                        .dataFetcher(fetcher(
                                GitChangeLog.class,
                                changeLog -> gitService.getChangeLogCommits(changeLog).getCommits()
                        ))
                )
                // TODO Issues
                // TODO File changes
                // OK
                .build();
    }

}

File: ontrack-extension-git/src/main/java/net/nemerosa/ontrack/extension/git/graphql/GitChangeLogGQLType.kt
package net.nemerosa.ontrack.extension.git.graphql

import graphql.Scalars.GraphQLString
import graphql.schema.DataFetchingEnvironment
import graphql.schema.GraphQLObjectType
import net.nemerosa.ontrack.extension.api.model.IssueChangeLogExportRequest
import net.nemerosa.ontrack.extension.git.model.GitChangeLog
import net.nemerosa.ontrack.extension.git.model.GitProjectNotConfiguredException
import net.nemerosa.ontrack.extension.git.service.GitService
import net.nemerosa.ontrack.graphql.schema.GQLType
import net.nemerosa.ontrack.graphql.schema.GQLTypeCache
import net.nemerosa.ontrack.graphql.support.listType
import org.springframework.http.HttpStatus
import org.springframework.http.ResponseEntity
import org.springframework.stereotype.Component

/**
 * GraphQL type for a change log
 * @see net.nemerosa.ontrack.extension.git.model.GitChangeLog
 */
@Component
class GitChangeLogGQLType(
    private val gitUICommitGQLType: GitUICommitGQLType,
    private val gitChangeLogIssuesGQLType: GitChangeLogIssuesGQLType,
    private val issueChangeLogExportRequestGQLInputType: IssueChangeLogExportRequestGQLInputType,
    private val gitService: GitService,
) : GQLType {

    override fun getTypeName(): String = GIT_CHANGE_LOG

    override fun createType(cache: GQLTypeCache): GraphQLObjectType {
        return GraphQLObjectType.newObject()
            .name(GIT_CHANGE_LOG)
            // Commits
            .field { f ->
                f.name("commits")
                    .description("List of commits in the change log")
                    .type(listType(gitUICommitGQLType.typeRef))
                    .dataFetcher { env ->
                        val gitChangeLog: GitChangeLog = env.getSource()
                        gitService.getChangeLogCommits(gitChangeLog).commits
                    }
            }
            // Issues
            .field { f ->
                f.name("issues")
                    .description("List of issues in the change log")
                    .type(gitChangeLogIssuesGQLType.typeRef)
                    .dataFetcher { env ->
                        val gitChangeLog: GitChangeLog = env.getSource()
                        gitService.getChangeLogIssues(gitChangeLog)
                    }
            }
            // Export of change log
            .field { f ->
                f.name("export")
                    .description("Export of the change log according to some specifications")
                    .type(GraphQLString)
                    .argument {
                        it.name("request")
                            .description("Export specifications")
                            .type(issueChangeLogExportRequestGQLInputType.typeRef)
                    }
                    .dataFetcher { env ->
                        val gitChangeLog: GitChangeLog = env.getSource()
                        // Parses the request
                        val request = parseExportRequest(env)
                        // Build boundaries
                        request.from = gitChangeLog.from.build.id
                        request.to = gitChangeLog.to.build.id
                        // Gets the associated project
                        val project = gitChangeLog.project
                        // Gets the configuration for the project
                        val gitConfiguration = gitService.getProjectConfiguration(project)
                            ?: return@dataFetcher null
                        // Gets the issue service
                        val optConfiguredIssueService = gitConfiguration.configuredIssueService
                        if (!optConfiguredIssueService.isPresent) {
                            return@dataFetcher null
                        }
                        val configuredIssueService = optConfiguredIssueService.get()
                        // Gets the issue change log
                        val changeLogIssues = gitService.getChangeLogIssues(gitChangeLog)
                        // List of issues
                        val issues = changeLogIssues.list.map { it.issue }
                        // Exports the change log using the given format
                        val exportedChangeLogIssues = configuredIssueService.issueServiceExtension
                            .exportIssues(
                                configuredIssueService.issueServiceConfiguration,
                                issues,
                                request
                            )
                        // Returns the content
                        exportedChangeLogIssues.content
                    }
            }
            // OK
            .build()
    }

    private fun parseExportRequest(env: DataFetchingEnvironment): IssueChangeLogExportRequest {
        val requestArg: Any? = env.getArgument<Any>("request")
        return issueChangeLogExportRequestGQLInputType.convert(requestArg) ?: IssueChangeLogExportRequest()
    }

    companion object {
        const val GIT_CHANGE_LOG = "GitChangeLog"
    }
}